src/chunkserver/io_buffers.cc

```diff
@@ -0,0 +1,438 @@
+	}
+}
+
+InputBuffer::InputBuffer(size_t headerSize, size_t numBlocks)
+	: headerSize_(headerSize),
+	  numBlocks_(numBlocks),
+	  blockBuffer_(numBlocks * SFSBLOCKSIZE, disk::kIoBlockSize),
+	  headerBuffer_(numBlocks * headerSize) {}
+
+ssize_t InputBuffer::readFromSocket(int sock, size_t bytesToRead) {
+	ssize_t bytesRead = 0;
+	auto bytesReadInHeaderBuffer = headerBuffer_.totalBytesPutInBuffer();
+	auto targetBytesInHeaderBuffer = writeInfo_.size() * headerSize_;
+
+	if (bytesReadInHeaderBuffer < targetBytesInHeaderBuffer) {
+		// If the header buffer is not filled, we read to the header buffer.
+		bytesRead = headerBuffer_.readFromFD(
+		    sock, std::min(targetBytesInHeaderBuffer - bytesReadInHeaderBuffer, bytesToRead));
+
+		if (bytesRead <= 0) { return bytesRead; }
+	}
+	bytesReadInHeaderBuffer = headerBuffer_.totalBytesPutInBuffer();
+
+	if (bytesReadInHeaderBuffer == targetBytesInHeaderBuffer &&
+	    static_cast<size_t>(bytesRead) < bytesToRead) {
+		// If the header buffer is already filled, we would need to read to the block buffer.
+		auto blockBytesRead = blockBuffer_.readFromFD(sock, bytesToRead - bytesRead);
+
+		if (blockBytesRead <= 0) {
+			return bytesRead;  // Return the number of bytes read so far.
+		}
+
+		bytesRead += blockBytesRead;
+	}
+	return bytesRead;
+}
+
+ssize_t InputBuffer::writeToSocket(int sock, size_t bytesToWrite) {
+	ssize_t bytesWritten = 0;
+	size_t bytesInHeaderBuffer = headerBuffer_.bytesInABuffer();
+	if (bytesInHeaderBuffer > 0) {
+		// If the header buffer is not flushed, we write from the header buffer.
+		size_t headerBytesToWrite = std::min(bytesToWrite, bytesInHeaderBuffer);
+		bytesWritten = headerBuffer_.writeToFD(sock, headerBytesToWrite);
+
+		if (bytesWritten <= 0) { return bytesWritten; }
+	}
+	bytesInHeaderBuffer = headerBuffer_.bytesInABuffer();
+
+	if (bytesInHeaderBuffer == 0) {
+		// If the header buffer is already flushed, we would need to write from the block buffer.
+		auto blockBytesWritten = blockBuffer_.writeToFD(sock, bytesToWrite - bytesWritten);
+
+		if (blockBytesWritten <= 0) {
+			return bytesWritten;  // Return the number of bytes written so far.
+		}
+
+		bytesWritten += blockBytesWritten;
+	}
+	return bytesWritten;
+}
+
+ssize_t InputBuffer::copyIntoBuffer(BufferType type, const void *mem, size_t len) {
+	switch (type) {
+	case BufferType::Block:
+		return blockBuffer_.copyIntoBuffer(mem, len);
+	case BufferType::Header:
+		return headerBuffer_.copyIntoBuffer(mem, len);
+	default:
+		safs::log_warn("(InputBuffer) Invalid buffer type");
+		return 0;
+	}
+}
+
+const uint8_t *InputBuffer::rawData(BufferType type) const {
+	switch (type) {
+	case BufferType::Block:
+		return blockBuffer_.paddedIndex(0);
+	case BufferType::Header:
+		return headerBuffer_.paddedIndex(0);
+	default:
+		safs::log_warn("(InputBuffer) Invalid buffer type");
+		return 0;
+	}
+}
+
+const uint8_t *InputBuffer::getStartLastWriteOperationHeader() {
+	if (writeInfo_.empty()) {
+		safs::log_warn(
```

Comment:
Consider using `__func__` macro to write the name of the function.
Extend to the rest of the code using it.
```suggestion
		safs::log_warn("{} called without any write operations.", __func__);
```